src/librustc_typeck/check/_match.rs

```diff
@@ -378,12 +378,61 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             // `fn foo(foo: &u32)`
                             if let Some(mut err) = err {
                                 if let PatKind::Binding(..) = inner.node {
-                                    if let Ok(snippet) = tcx.sess.source_map()
-                                                                    .span_to_snippet(pat.span)
-                                    {
-                                        err.help(&format!("did you mean `{}: &{}`?",
-                                                            &snippet[1..],
-                                                            expected));
+                                    let parent_id = tcx.hir().get_parent_node_by_hir_id(pat.hir_id);
+                                    let parent = tcx.hir().get_by_hir_id(parent_id);
+                                    debug!("inner {:?} pat {:?} parent {:?}", inner, pat, parent);
+                                    match parent {
+                                        hir::Node::Item(hir::Item {
+                                            node: hir::ItemKind::Fn(..), ..
+                                        }) |
+                                        hir::Node::ForeignItem(hir::ForeignItem {
+                                            node: hir::ForeignItemKind::Fn(..), ..
+                                        }) |
+                                        hir::Node::TraitItem(hir::TraitItem {
+                                            node: hir::TraitItemKind::Method(..), ..
+                                        }) |
+                                        hir::Node::ImplItem(hir::ImplItem {
+                                            node: hir::ImplItemKind::Method(..), ..
+                                        }) => { // this pat is likely an argument
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(inner.span)
+                                            { // FIXME: turn into structured suggestion, will need
+                                              // a span that also includes the the arg's type.
+                                                err.help(&format!(
+                                                    "did you mean `{}: &{}`?",
+                                                    snippet,
+                                                    expected,
+                                                ));
+                                            }
+                                        }
+                                        hir::Node::Expr(hir::Expr {
+                                            node: hir::ExprKind::Match(..), ..
```

Comment:
Presumably this will also impact `if let`... if so, can you extend the tests below to account for that?

Reply:
`if let` desugars to `hir::ExprKind::Match`, so it is handled here.

Reply:
Sure; [I know](https://github.com/rust-lang/rust/pull/59288) ;) -- but I didn't see any test that exercised the behavior for `if let`. I think it's not a good testing methodology to assume the implementation strategy (e.g. that `if let` is desugared to `match`) in your tests. It is entirely possible that `if let` stops being desugared to `match` one day if e.g. we have `hir::ExprKind::Let` + `hir::ExprKind::If`.

Reply:
I read too quickly and I thought you meant it wasn't handled, my bad ^_^

Added in last commit.

Reply:
Hehe =P Thanks!